### DISCUSSION TITLE
Bluetooth Multimeter Problems

### DISCUSSION THREAD

**Discussion by Tomsdasd** [webspiderteam/Bluetooth-DMM-For-Windows#42] [General]
Hello there,
I've recently been using a Bluetooth multimeter and I'm trying to connect it to my computer to obtain the readings. I've managed to establish the Bluetooth connection, but I'm unsure how to retrieve the data. I've attempted to use your provided code, but it doesn't seem to work. Could you please help me with a solution? Thank you very much!

**Comment by ludwich66** [comment]
Hello,
nice that you want to use the software.
Could you please be more specific, what device are you using?
Did the BT pairing work?
Have you looked at the wiki pages?
We'll figure it out ;-)

**Comment by Tomsdasd** [comment]
Currently, I am using the ZT-5BQ device, which connects via Bluetooth Low Energy. I have seen the Bluetooth analysis page, but I'm not sure how to use it.

**Reply by ludwich66** [reply]
What do you expect from the program?
What do you want to do?
???
"You're not sure how to use this"
???
1. BT Connect
2. View measured values / see progression over time
3. Remote control BT-DMM from PC
4. Save data / distribute via MQTT

I'm happy to help, but it would be nice if there was a description of the problem, not just three words...

I want... how can I achieve this.
I have already tried the following...

**Reply by Tomsdasd** [reply]
I want to use Bluetooth to receive the current readings detected by ZT-5BQ. I have successfully connected via Bluetooth, and now I want to view the measured values. However, I am unsure how to receive the measurements. I tried using UUIDs and characteristic values to find the data, but I'm not sure if I am using the correct ones.

**Reply by ludwich66** [reply]
So you don't use the "ready-made" Windows program but would like to build something "own" on the basis?
If so, we have to wait for the programmer Engin, I'm the part that created the wiki ;-)

Did you find this Infos?  https://github.com/ludwich66/Bluetooth-DMM

**Reply by webspiderteam** [reply]
Hello. First of all which programing language did you used? 
BTW why you are trying to build it? is there a problem in app?

**Reply by Tomsdasd** [reply]
@ludwich66 Thanks you! I'm glad you found this information.

@webspiderteam Hi I am doing bluetooth programming in C#. Recently, I've been learning how to program bluetooth applications, and by coincidence, I have this tool (ZT-5BQ) handy. I'd like to research and experiment with it, but I'm having trouble retrieving the measurements. I couldn't find information on this, so I came here for help.
The app itself works fine, but when I download the source code and try to execute it, it fails and I'm not sure what the problem is.

**Reply by webspiderteam** [reply]
UUID's are "0000fff0-0000-1000-8000-00805f9b34fb" Service UUID
"0000fff4-0000-1000-8000-00805f9b34fb" Characteristic UUID 
but datas are not simple for these devices. all data is encodet bit by bit . you can use informations on ludwich's page https://github.com/ludwich66/Bluetooth-DMM all data protocols are there. 
> when I download the source code and try to execute it, it fails and I'm not sure what the problem is.

there is some settings. first of all you have to use .net framework 4.7.2 or lower version because it gives errors somehow. 
the second think is windows referances is problem visual studio connot add windows referances. so you have to add it manually. In referances you will see some referances has triange signs you have to remove them and then you have to add this referance "C:\Program Files (x86)\Windows Kits\10\UnionMetadata\10.0.16299.0\Windows.winmd" file path is normally that. if you dont have this version you can try others but I dont know if its works correctly. after these right clik on solution and click on Clean Solution then it shoul work without problem.

**Reply by Tomsdasd** [reply]
Understood. Tomorrow, I will test adding the Windows reference. However, if I simply want to retrieve a single measurement value or other data, do I use UUIDs and characteristic values to find the data, or do I need other methods?

**Reply by webspiderteam** [reply]
Yes you can get data but as I said it's not simple data to view directly measurements you have to make an encoding code the retrieve measurements as seen on multimeter

**Reply by Tomsdasd** [reply]
Understood. However, I would like to ask about the process of writing the decoding program. Is there any reference material available for me to consult?

**Reply by webspiderteam** [reply]
all process is explained in ludwichs page https://github.com/ludwich66/Bluetooth-DMM 
but basically you have to make Xor operation with these hex values


41 | 21 | 73 | 55 | a2 | c1 | 32 | 71 | 66 | aa | 3b | d0 | e2 | a8 | 33 | 14 | 20 | 21 | aa | bb
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --



for each byte. for example first byte  b1^41 second byte b2^21 ....
after that you will get clean results for exampla if auto range symbol on you will get 1 at bit 87 (for big dmms) other wise it can be complicated. because before xor some of them will be 0 for on state others will be 1.  after these you have to find which bit makes what but as I said before we already published all bit values in ludwichs page you can find all in there. the other prombel is the digits. if you wil not use a someting like 7seqment display (that every part of digit responds a bit) then you have to get 7 bits of data for a digit and change with equal value (for example 1111101 can be 0  or 0100010 can be means 1. I selected these values randomly for giving example). this can be your starting point.

**Reply by Tomsdasd** [reply]
Understood, thank you so much for your help. I have made progress on my project in the past few days. However, I would like to inquire about the line chart part. Are there any related packages available for use in UWP to create line charts?

**Reply by webspiderteam** [reply]
I had found while working with up but it was long time ago. I can't remember what it was

**Reply by Tomsdasd** [reply]
Understood, thank you very much for your response!

**Comment by webspiderteam** [comment]
> I want to use Bluetooth to receive the current readings detected by ZT-5BQ. I have successfully connected via Bluetooth, and now I want to view the measured values. However, I am unsure how to receive the measurements. I tried using UUIDs and characteristic values to find the data, but I'm not sure if I am using the correct ones.

"0000fff0-0000-1000-8000-00805f9b34fb" Service UUID
"0000fff4-0000-1000-8000-00805f9b34fb" Characteristic UUID
